### PR TITLE
MVP release

### PR DESCRIPTION
**Pull request by kevinfodness**
* Add ability to configure feature flags via filter
* Add ability to turn feature flags on or off via a Settings page
* Add ability to get value of a feature flag via filter
* Add tests and Travis config
* Add docs

### CODE REVIEW

File: phpcs.xml.dist [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    <arg name="extensions" value="php" />
+
+    <!-- Set severity to 1 to see everything that isn't effectively turned off. -->
+    <arg name="severity" value="1" />
+
+    <!-- Check all files in this directory and the directories below it. -->
+    <file>.</file>
+
+    <rule ref="WordPress" />
+    <rule ref="WordPress-VIP-Go" />
```
Comment (joemcgill) [inline phpcs.xml.dist:31]: The Alley-Interactive ruleset includes both of these plus our own preferences, like array syntax.

**Review by joemcgill — APPROVED** [review]
This is a nice implementation! I'm curious about your decision to explicitly use the filters API explicitly for this functionality, rather than wrapping it in an API that is specific to this functionality, e.g., `Experimental_Features\register_feature()` and `Experimental_Features\is_active()`, etc.

Left several suggestions, but no real blockers. Nice work!

**Comment by kevinfodness** [comment]
Thanks for the review, @joemcgill - lots of good comments here. I'll work through resolving them shortly. As for your question about why I've chosen to use filters rather than functions, it's because I didn't want to create a hard dependency between a theme or a plugin and this plugin. You should be able to deactivate or uninstall this plugin without breaking anything in themes or plugins that are using it. Alternately, I could have recommended `$is_active = function_exists( 'experimental_features_get_flag' ) ? experimental_features_get_flag( 'flag-slug' ) : false` but that relies on theme or plugin authors implementing the function check. Using the filter system is a kind of dependency injection - one or more themes and plugins can use the filter to register their own feature flags, and then they can safely check for feature flags that are enabled via filter (which could be wrapped in a convenience function in an individual theme or plugin).

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  - export PATH="$HOME/.config/composer/vendor/bin:$PATH"
+
+  # Install phpunit.
+  - composer global require "phpunit/phpunit=6.5.*"
+
+  # Set up WordPress installation.
+  - bash bin/install-wp-tests.sh wordpress_test root '' localhost $WP_VERSION
+
+  # Maybe set up phpcs.
+  - |
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply .travis.yml:59]: Yep, I'll wire that up right now.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,49 @@
...
+### Defining Feature Flags
+
+Feature flags are defined using a filter:
+
+```php
+/**
+ * Define available feature flags.
+ *
+ * @param array $flags Feature flags that have been defined for the Experimental Features plugin.
+ *
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply README.md:16]: 👍

File: README.md [hunk 4]
```diff
@@ -1,2 +1,49 @@
...
+ * @return array The modified list of feature flags.
+ */
+function filter_experimental_features_flags( array $flags ) : array {
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply README.md:19]: 👍

File: README.md [hunk 5]
```diff
@@ -1,2 +1,49 @@
...
+	$flags['my-cool-feature'] = __( 'My Cool Feature', 'my-textdomain' );
+
+	return $flags;
+}
+add_filter(
+	'experimental_features_flags',
+	__NAMESPACE__ . '\filter_experimental_features_flags',
+	10,
+	1
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply README.md:28]: 👍

File: bin/install-wp-tests.sh [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+	if [ -d $WP_CORE_DIR ]; then
+		return;
+	fi
+
+	mkdir -p $WP_CORE_DIR
+
+	if [[ $WP_VERSION == 'nightly' || $WP_VERSION == 'trunk' ]]; then
+		mkdir -p $TMPDIR/wordpress-nightly
+		download https://wordpress.org/nightly-builds/wordpress-latest.zip  $TMPDIR/wordpress-nightly/wordpress-nightly.zip
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply bin/install-wp-tests.sh:63]: I snagged this from `wp-starter-theme` - looks like we will need to update it there, also.

File: inc/class-filter.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	 *
+	 * @return bool Whether the given feature flag is active or not.
+	 */
+	public static function filter_experimental_features_flag( bool $default, string $slug ) : bool {
+		$option_value = get_option( 'experimental_features_flags', [] );
+
+		return ! empty( $option_value )
+			&& is_array( $option_value )
+			&& in_array( $slug, $option_value, true )
+				? true
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply inc/class-filter.php:47]: Enabled flags are stored as slugs in an array. So the actual value of the option with a slug called `my-cool-feature` would be `[ 'my-cool-feature' ]`.

File: inc/class-options.php [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+				'label'    => esc_html__( 'Feature Flags', 'experimental-features' ),
+				'sanitize' => [ self::class, 'sanitize_flags' ],
+			],
+		];
+
+		// Register the config section.
+		add_settings_section(
+			'experimental_features_config_section',
+			esc_html__( 'Experimental Features Configuration', 'experimental-features' ),
+			null,
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply inc/class-options.php:32]: 👍

File: inc/class-options.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	/**
+	 * Adds the options page to the Settings menu.
+	 */
+	public static function action_admin_menu() {
+		add_options_page(
+			esc_html__( 'Experimental Features', 'experimental-features' ),
+			esc_html__( 'Experimental Features', 'experimental-features' ),
+			Filter::capability(),
```
[the remark this answers is not in the document]
Comment (kevinfodness) [reply inc/class-options.php:68]: `Filter` here is in the same namespace and is loaded via the same autoloader that this class is. Is it still a best practice to include it via a `use` statement in that circumstance? What is the advantage there? I'm happy to make the change if it's a best practice.

File: phpcs.xml.dist (hunk 1 above)
Comment (kevinfodness) [reply phpcs.xml.dist:31]: 👍

File: inc/class-options.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (joemcgill) [reply inc/class-options.php:68]: No, you're probably right. I tend to error on the side of overuse to make it easy to see where external classes are coming from. There's not functional reason to add them as far as I can see.

**Pull request merged**